python/ray/_private/external_storage.py

```diff
@@ -276,7 +276,10 @@ def __init__(self, directory_path, buffer_size=None):
 
         # Create directories.
         for path in directory_path:
-            full_dir_path = os.path.join(path, self._spill_dir_name)
+            if node_id:
+                full_dir_path = os.path.join(path, node_id, self._spill_dir_name)
```

Comment:
@fishbone should we actually do 

```
os.path.join(path, self._spill_dir_name, node_id)
```

Reply:
Maybe we can just do `os.path.join(path, f"{self._spill_dir_name}_{node_id}")`

Reply:
@jjyao Maybe not `os.path.join(path, f"{self._spill_dir_name}_{node_id}")`

Two things:
- `_spill_dir_name` might end up with `/` which needs to be taken care of
- The user assume the data will be put into `_spill_dir_name`, not `{_spill_dir_name}_{node_id}`

But I think it's a good catch.


@ruisearch42 let's do `os.path.join(path, self._spill_dir_name, node_id)` Basically we put everything inside the dir the user asks.

Reply:
Currently _spill_dir_name hard coded to "ray_spilled_objects", let's say path is "/tmp/spill", the question is if we set full_dir_path to /tmp/spill/<node_id>/ray_spilled_objects or /tmp/spill/ray_spilled_objects/<node_id>

Since in `destroy_external_storage()`, the code currently does rmtree for each full_dir_path, we would be left with either /tmp/spill/<node_id> or /tmp/spill/ray_spilled_objects which fails unit tests. We need to delete the parent directory as well, which is not clean code (either we need to delete parent directory which makes assumptions about the directory structure, or we introduce "clean_up_dir_path"s).

I'm thinking of either using `os.path.join(path, f"{self._spill_dir_name}_{node_id}")` as Jiajun mentioned (since _spill_dir_name is currently hard coded to "ray_spilled_objects"), or the following:
```
            if node_id:
                full_dir_path = os.path.join(path, node_id)
            else:
                full_dir_path = os.path.join(path, self._spill_dir_name)
```

Thoughts? @fishbone @jjyao

Reply:
Yes, that's why I suggest ray_spilled_objects_{node_id}. It's the simplest, anyway ray_spilled_object is already a private folder we create under users folder.

Reply:
`DEFAULT_OBJECT_PREFIX` is defined in python/ray/_private/ray_constants.py:
DEFAULT_OBJECT_PREFIX = "ray_spilled_objects"

This is considered as private API and subject to change?

Reply:
Yes I think so. I don't see it's mentioned in the public doc.

cc @fishbone @rkooo567 to double confirm.

Reply:
@fishbone I was also rethinking the original ask/proposal: since spilled objects have distinct file names (uuid), how would they collide?

Reply:
it is not public

Reply:
@ruisearch42 one issue might be that when one node exits, it will delete the folder which is still used by others?

Reply:
I see. If it's just in the dir provided by the user, I'm OK with either way.